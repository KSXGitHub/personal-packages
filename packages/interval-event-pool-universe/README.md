# Interval Event Pool Universe

A pool of interval event pool

## Usage

```javascript
const { INTERVAL_EVENT, getEventPool } = require('@khai96x/interval-event-pool-universe')

const target = getEventPool(1000)
  .addListener(INTERVAL_EVENT, function listener (param) {
    if (param.intervalCount === 10) {
      target.removeListener(INTERVAL_EVENT, listener)
      target.stopEventLoop()
    } else {
      console.log(param)
    }
  })
```

**Expected Output:**

```
IntervalEventParam { intervalDelay: 1000, intervalCount: 1 }
IntervalEventParam { intervalDelay: 1000, intervalCount: 2 }
IntervalEventParam { intervalDelay: 1000, intervalCount: 3 }
IntervalEventParam { intervalDelay: 1000, intervalCount: 4 }
IntervalEventParam { intervalDelay: 1000, intervalCount: 5 }
IntervalEventParam { intervalDelay: 1000, intervalCount: 6 }
IntervalEventParam { intervalDelay: 1000, intervalCount: 7 }
IntervalEventParam { intervalDelay: 1000, intervalCount: 8 }
IntervalEventParam { intervalDelay: 1000, intervalCount: 9 }
```

## Notes

1. Event loop will start immediately after its creation

2. Providing the same delay value returns the same event pool, and providing different delay values gives different event pool

```javascript
assert.equal(
  getEventPool(1000),
  getEventPool(1000)
)

assert.notEqual(
  getEventPool(1000),
  getEventPool(1001)
)
```

## License

[MIT](https://git.io/fj9XO) © [Hoàng Văn Khải](https://github.com/KSXGitHub)
