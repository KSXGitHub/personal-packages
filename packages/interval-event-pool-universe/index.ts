import { createEventPool, some } from '@khai96x/simple-event-pool'
import AdvMapInit from 'advanced-map-initialized'

export const INTERVAL_EVENT = Symbol('INTERVAL_EVENT')

const universe = new AdvMapInit(
  Map,
  (delay: number) => createEventPool({
    setInterval,
    clearInterval,
    delay
  })
    .createAutoTrigger(INTERVAL_EVENT, param => {
      return some(new IntervalEventParam(
        delay,
        param.iterationCount
      ))
    })
)

class IntervalEventParam {
  constructor (
    public readonly internalDelay: number,
    public readonly internalCount: number
  ) {}
}

export function get (delay: number) {
  return universe.get(delay)
}
