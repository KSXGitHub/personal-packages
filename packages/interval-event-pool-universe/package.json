{
  "name": "@khai96x/interval-event-pool-universe",
  "version": "0.0.0",
  "description": "A pool of interval event pool",
  "author": "Hoàng Văn Khải <hvksmr1996@gmail.com>",
  "license": "MIT",
  "homepage": "https://github.com/KSXGitHub/personal-packages#readme",
  "repository": {
    "type": "git",
    "url": "git+https://github.com/KSXGitHub/personal-packages.git"
  },
  "bugs": {
    "url": "https://github.com/KSXGitHub/personal-packages/issues"
  },
  "dependencies": {
    "@khai96x/simple-event-pool": "^0.0.3",
    "advanced-map-initialized": "^0.0.2",
    "tslib": "^1.10.0",
    "@types/node": "^12.6.8"
  }
}
