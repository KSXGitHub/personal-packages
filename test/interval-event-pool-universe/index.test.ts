import createLock from 'remote-controlled-promise'
import { INTERVAL_EVENT, IntervalEventParam, getEventPool } from '@khai96x/interval-event-pool-universe'

describe('getEventPool ： delay → event pool', () => {
  it('event pool of the same delay is the same', () => {
    expect(getEventPool(10)).toBe(getEventPool(10))
  })

  it('event pool of different delay is different', () => {
    expect(getEventPool(10)).not.toBe(getEventPool(12))
  })
})

describe('interval event', () => {
  function setup () {
    const DELAY = 3
    const LOCK_COUNT = 10

    const pool = getEventPool(DELAY)
      .addListener(INTERVAL_EVENT, function listener (param) {
        if (param.intervalCount > LOCK_COUNT) {
          pool.removeListener(INTERVAL_EVENT, listener)
        } else {
          void lockArray[param.intervalCount - 1].resolve(param)
        }
      })

    const lockArray = Array(LOCK_COUNT)
      .fill(undefined)
      .map(() => createLock<IntervalEventParam>())

    const lockPromise = Promise.all(lockArray.map(lock => lock.promise))

    return { DELAY, LOCK_COUNT, pool, lockArray, lockPromise }
  }

  it('with expected arguments', async () => {
    const { lockPromise } = setup()
    expect(await lockPromise).toMatchSnapshot()
  })

  describe('every argument has "intervalDelay" property', () => {
    const { DELAY, lockArray } = setup()

    lockArray.forEach((lock, index) => {
      it(`argument #${index}`, async () => {
        expect(await lock.promise).toHaveProperty('intervalDelay', DELAY)
      })
    })
  })

  describe('every argument has "intervalCount" property', () => {
    const { lockArray } = setup()

    lockArray.forEach((lock, index) => {
      it(`argument #${index}`, async () => {
        expect(await lock.promise).toHaveProperty('intervalCount', expect.any(Number))
      })
    })
  })
})
