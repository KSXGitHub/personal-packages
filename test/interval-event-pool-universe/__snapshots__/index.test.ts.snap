// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`interval event with expected arguments 1`] = `
Array [
  IntervalEventParam {
    "intervalCount": 1,
    "intervalDelay": 3,
  },
  IntervalEventParam {
    "intervalCount": 2,
    "intervalDelay": 3,
  },
  IntervalEventParam {
    "intervalCount": 3,
    "intervalDelay": 3,
  },
  IntervalEventParam {
    "intervalCount": 4,
    "intervalDelay": 3,
  },
  IntervalEventParam {
    "intervalCount": 5,
    "intervalDelay": 3,
  },
  IntervalEventParam {
    "intervalCount": 6,
    "intervalDelay": 3,
  },
  IntervalEventParam {
    "intervalCount": 7,
    "intervalDelay": 3,
  },
  IntervalEventParam {
    "intervalCount": 8,
    "intervalDelay": 3,
  },
  IntervalEventParam {
    "intervalCount": 9,
    "intervalDelay": 3,
  },
  IntervalEventParam {
    "intervalCount": 10,
    "intervalDelay": 3,
  },
]
`;
